core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/messaginglayer/NetworkSenderActor.scala

```diff
@@ -30,29 +45,31 @@ object NetworkOutputGate {
     * @param messageID
     */
   final case class NetworkAck(messageID: Long)
+
+  def props(): Props =
+    Props(new NetworkSenderActor())
 }
 
-/** This trait handles the transformation from identifier to actorRef
+/** This actor handles the transformation from identifier to actorRef
   * and also sends message to other actors. This is the most outer part of
   * the messaging layer.
-  * It SHOULD be a trait since it is highly coupled with actors' sending/receiving logic.
-  * TODO: this trait should eventually become another actor, which is the sender actor.
   */
-trait NetworkOutputGate {
-  this: Actor => // it requires the class to be an actor.
+class NetworkSenderActor extends Actor {
 
-  private val idToActorRefs = mutable.HashMap[ActorVirtualIdentity, ActorRef]()
-  private val messageStash = mutable.HashMap[ActorVirtualIdentity, mutable.Queue[WorkflowMessage]]()
+  val idToActorRefs = new mutable.HashMap[ActorVirtualIdentity, ActorRef]()
+  val messageStash = new mutable.HashMap[ActorVirtualIdentity, mutable.Queue[WorkflowMessage]]()
+  val idToCongestionControls = new mutable.HashMap[ActorVirtualIdentity, CongestionControl]()
 
   /** keeps track of every outgoing message.
     * Each message is identified by this monotonic increasing ID.
     * It's different from the sequence number and it will only
     * be used by the output gate.
     */
-  private var networkMessageID = 0L
+  var messageID = 0L
```

Comment:
I think renaming this to networkMessageID makes the code more readable

Reply:
renamed